Client print error when specify wrong detach keys

WeiZhang555 opened the pull request:
Fix #21064

Let client print error message explicitly when user specifies wrong
detach keys.

Sample output:

```
$ docker run --rm -ti  --detach-keys  "ctrl-D,ctrl-e" busybox sh
WARNING: Invalid escape keys provided (ctrl-D,ctrl-e) using default : ctrl-p ctrl-q
/ # 
```

Signed-off-by: Zhang Wei zhangwei555@huawei.com

WeiZhang555 commented:
I'll add a test case later

WeiZhang555 commented:
Test case is done.

WeiZhang555 commented:
It's funny...Janky test passed, but its status is always pending, the yellow dot is stuck there.
@thaJeztah Could you fix it?

thaJeztah commented:
@WeiZhang555 I think GitHub missed a notification then. No need to run the test again for now :-)

thaJeztah commented:
I wonder if this should just be a *warning*, IMO, producing error and aborting the "exec" sounds more logical; it's the same we do if I specify an unknown flag

vdemeester commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -196,6 +196,75 @@ func (s *DockerSuite) TestRunAttachDetachFromFlag(c *check.C) {
 }
 
+// TestRunDetach checks attaching and detaching with the escape sequence specified via flags.
+func (s *DockerSuite) TestRunAttachDetachFromInvalidFlag(c *check.C) {
```
Is it just a new integration test for checking the warning ? (if yes, why not using a existing one ?)

WeiZhang555 replied:
Not only. 
I can find two existing `--detach-key` tests
* TestRunAttachDetachKeysOverrideConfig: checks attaching and detaching with the detach flags, making sure it overrides config file.
* TestRunDetach: checks attaching and detaching with the detach flags, making sure it overrides config file

What this test is doing is: input an invalid detach key, it will throw a warning, then fall back to default detach keys, make sure default key works properly. 

Currently we don't have an integration test for testing default detach key as I can see, so this will also be a good supplement.

WeiZhang555 replied:
And it looks like `TestRunAttachDetachKeysOverrideConfig` can cover situations in `TestRunDetach`, but that's another problem

WeiZhang555 commented:
@thaJeztah 

> I wonder if this should just be a warning, IMO, producing error and aborting the "exec" sounds more logical; it's the same we do if I specify an unknown flag

Yes, this is a good question, currently on the daemon side this is only a warning, so I leave it as a warning to user instead of a fatal error. Besides that, this won't be first invalid option which will only print an warning, see:
```
$ docker run -ti --memory 100M busybox sh
WARNING: Your kernel does not support swap limit capabilities, memory limited without swap.
/ #
```
We already have others, so I can't say which one is correct as we don't have a policy of how to handle invalid flags(maybe we have?).

thaJeztah commented:
> We already have others, so I can't say which one is correct as we don't have a policy of how to handle invalid flags

You're right, we don't have a policy (I think we should have one). Personally I'd be really in favor of failing; I think it'd be better, and consistent with "invalid" flags.

One question (I didn't look too closely), with this PR, do we still have daemon-side validation, or only client-side? I glanced quickly and saw some checks were removed.


@vdemeester wdyt on error vs warning?

WeiZhang555 commented:
@thaJeztah 

I thought about it again, and realized that the example of memory limit I gave is not the same with `--detach-keys`. Even without swap limit capabilities, the memory limit can still be done, so it's appropriate to be a warning.
Now I'm strong +1 on your idea, this should be a fatal error and should fail.

>One question (I didn't look too closely), with this PR, do we still have daemon-side validation, or only client-side? I glanced quickly and saw some checks were removed.

Don't worry, the validation is still on daemon side, I'm just moving some duplicate code into a deeper function call. :smile:

vdemeester commented:
@thaJeztah @WeiZhang555 Let's be naive and erroring out if the keys are not correct :stuck_out_tongue:. It feels a little overkill to fail the start of the container for this, but on the other hand, it's an invalid input so… it's an error. 

So I'm +1 too on erroring out.

calavera commented:
LGTM

calavera commented:
fwiw janky's tests passed but GitHub didn't get the notification:

```
04:13:14 + docker run --rm -v /home/ubuntu/workspace/Docker-PRs:/workspace busybox chown -R 1000:1000 /workspace
04:13:16 POST BUILD TASK : SUCCESS
04:13:16 END OF POST BUILD TASK : 0
04:13:16 Archiving artifacts
04:13:21 [PostBuildScript] - Execution post build scripts.
04:13:21 [Docker-PRs] $ /bin/sh -xe /tmp/hudson997090171588085758.sh
04:13:21 + set +x
04:13:21 
04:13:21 
04:13:21 ---
04:13:21 Now starting POST-BUILD steps
04:13:21 ---
04:13:21 
04:13:23 Failed to remove container (docker-pr24941): Error response from daemon: No such container: docker-pr24941
04:13:23 Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
04:13:23 Finished: SUCCESS
```

Consider it all :green_heart:

thaJeztah commented on api/client/attach.go:
```diff
@@ -42,6 +43,11 @@ func (cli *DockerCli) CmdAttach(args ...string) error {
+		// validate detach key
+		_, err := term.ToBytes(*detachKeys)
+		if err != nil {
+			return fmt.Errorf("Invalid escape keys provided (%s) using default : ctrl-p ctrl-q", *detachKeys)
```
The "using default..." is no longer needed now that we error out?

WeiZhang555 replied:
Ah, you're definitely right!

thaJeztah commented on daemon/attach.go:
```diff
@@ -11,10 +11,20 @@ import (
+	if c.DetachKeys != "" {
+		keys, err = term.ToBytes(c.DetachKeys)
+		if err != nil {
+			return fmt.Errorf("Invalid escape keys provided (%s) using default : ctrl-p ctrl-q", c.DetachKeys)
```
Same here?

WeiZhang555 commented:
I made some updates, please take a look.

vdemeester commented:
@WeiZhang555 I have a weird behaviour with the current implementation. If there is an error, it hangs (and there is no `WARNING:` in front of the message :stuck_out_tongue_closed_eyes:)

```bash
$ docker run --rm -ti  --detach-keys  "ctrl-D,ctrl-e" busybox sh
Invalid escape keys (ctrl-D,ctrl-e) provided
# and then.. it hangs.. no prompt, nothing..

# From another terminal 
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS              PORTS               NAMES
62016f02102d        busybox             "sh"                About a minute ago   Up About a minute                       serene_ritchie
```

WeiZhang555 commented:
@vdemeester You must be using an existing docker client with this daemon, sorry that I'm a little busy recent days and forgot to post more updates. I found an existing bug here and am trying to fix it in this PR.

in `api/client/attach.go`,  try to annotate these lines:
```
 33     /*
 34         if !c.State.Running {
 35             return fmt.Errorf("You cannot attach to a stopped container, start it first")
 36         }
 37 
 38         if c.State.Paused {
 39             return fmt.Errorf("You cannot attach to a paused container, unpause it first")
 40         }*/
```

If you annotate these if assertion and try to use this client to attach to a stopped or paused container, it will also hang. I think it's caused by a non-closed http connection. I already have some thoughts about it but am trying to find if there is better way.

Will post more updates later.

vdemeester commented:
ping @WeiZhang555 what is the status here ?😇

WeiZhang555 commented:
@vdemeester post an update, sorry for the late. 

I add some validation code on client side, you can try to remove it to see how it performs only on daemon side, it still have some tiny format problem, like this:
```
$ docker attach --detach-keys "ctrl-A,a" 385cfbf0c5ec
Error: Invalid escape keys (ctrl-A,a) provided
                                              $
```

This is existing problem:
```
$ docker attach --detach-keys "ctrl-A,a" 385cfbf0c5ec
Error: Invalid escape keys (ctrl-A,a) provided
                                              $
```

So one problem brings another problem, :sob: 

I want to handle this format problem in another PR if you can accept, or I can put fix into this one if you have any suggesion. :smile:

WeiZhang555 commented:
I break the http response status code, fixing now...

WeiZhang555 committed:
Let client print error when speicify wrong detach keys

Fix #21064

Let client print error message explicitly when user specifies wrong
detach keys.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
Hmmm, I added some logic to make sure the network connection is closed after HTTP error is sent. It works. As I said, before this, if you try to attach a paused container without client side validation, it hangs, actually it you don't add validation logic on client side, try to attach to a container, daemon will send an error and client will print this error then hangs.

So this commit can fix this. 

But I met a strange problem, so if I don't do the client side validation(or using an old client + this new daemon), it will print:
```
$ docker attach --detach-keys "ctrl-A,a" 7e7c5b6885b
28
Invalid escape keys (ctrl-A,a) provided

$
```
What is the 28? It looks like string length of "Invalid escape keys (ctrl-A,a) provided" but I can't see which part print this.
But checking the api with:
```
$ curl -v -X POST http://localhost:8888/v1.23/containers/7e7c5b6885/attach?detachKeys=ctrl-A%2Ca\&stderr=1\&stdin=1\&stdout=1\&stream=1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> POST /v1.23/containers/7e7c5b6885/attach?detachKeys=ctrl-A%2Ca&stderr=1&stdin=1&stdout=1&stream=1 HTTP/1.1
> User-Agent: curl/7.38.0
> Host: localhost:8888
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Content-Type: text/plain; charset=utf-8
* Server Docker/1.11.0-dev (linux) is not blacklisted
< Server: Docker/1.11.0-dev (linux)
< X-Content-Type-Options: nosniff
< Date: Thu, 24 Mar 2016 07:35:12 GMT
< Transfer-Encoding: chunked
< 
Invalid escape keys (ctrl-A,a) provided
* transfer closed with outstanding read data remaining
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

So apparently daemon doesn't send a string "28", do you have any suggestion? @vdemeester

WeiZhang555 commented:
Everything works now, please take a look.  @vdemeester @calavera 
Server will shutdown HTTP connection after sending error to client, in case client hangs.

coolljt0725 commented on daemon/attach.go:
```diff
@@ -11,10 +11,21 @@ import (
+		keys, err = term.ToBytes(c.DetachKeys)
+		if err != nil {
+			err = fmt.Errorf("Invalid escape keys (%s) provided", c.DetachKeys)
+			return err
```
how about `return fmt.Errorf("Invalid escape keys (%s) provided", c.DetachKeys)`  here?

WeiZhang555 replied:
Nice. :+1:

coolljt0725 commented on api/server/router/container/container_routes.go:
```diff
@@ -465,11 +456,24 @@ func (s *containerRouter) postContainersAttach(ctx context.Context, w http.Respo
+		if errHijack == nil {
+			statusCode := httputils.GetHTTPErrorStatusCode(err)
+			statusText := http.StatusText(statusCode)
+			fmt.Fprintf(conn, "HTTP/1.1 %d %s\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n%s\r\n", statusCode, statusText, err.Error())
```
should we also need handle `upgrade` here just like we do at https://github.com/docker/docker/pull/21092/files#diff-207a413f4aeb453c56da19388fc9621aR439 ?

WeiZhang555 replied:
I think it's OK to ignore the upgrade header and return directly the error status code.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism

It won't bring problem if I understand it right.

WeiZhang555 commented:
Updated. Thanks @coolljt0725 for helping review and debug. :smile:

WeiZhang555 commented:
ping @vdemeester @calavera

vdemeester commented:
@WeiZhang555 Took a look, and there is still one case that "hangs", when `--rm` is prodived.

```bash
$ docker attach --detach-keys "ctrl-A,a"  c3d730852c83133693c8c4adc4cb6a7f1f9bd32127b6a9180a27056cb906f12f
Invalid escape keys (ctrl-A,a) provided
$ docker run -ti  --detach-keys "ctrl-D,ctrl-e" busybox sh
Invalid escape keys (ctrl-D,ctrl-e) provided
$ docker run --rm -ti  --detach-keys  "ctrl-D,ctrl-e" busybox sh
Invalid escape keys (ctrl-D,ctrl-e) provided
# […] hangs 
```

:angel:

WeiZhang555 commented:
@vdemeester Nice catch!

It hangs on ContainerWait: https://github.com/docker/docker/blob/master/api/client/run.go#L266

So I need to make some vendor update to throw an error in ContainerAttach. Some extra changes:
1.force container to remove when specify `--rm`. Because when invalid detach key is provided, container may start successfully somehow, `--rm` can only work when force container to remove.
2.Handle detach key error in start api. e.g.
```
$ docker start -ai --detach-keys "ctrl-A,ctrl-d" fecf527ff75541
Invalid escape keys (ctrl-A,ctrl-d) provided
persistent connection closed
```
3.Handle detach key error in `docker exec`:
```
$ docker exec -ti --detach-keys "ctrl-A,ctrl-e" fecf527ff755 sh
Error response from daemon: Invalid escape keys (ctrl-A,ctrl-e) provided
```

PTAL. :smile:

vdemeester commented:
@WeiZhang555 It's fixed :wink: You can make a PR in `docker/engine-api` then :wink:

WeiZhang555 commented:
@vdemeester OK, thank you for your so careful test :pray: :smile:

vdemeester commented:
@WeiZhang555 engine-api PR is merged, you can vendor it here and update the PR :wink:

WeiZhang555 commented:
All right, I'm on travel these days, will update in 3 or 4 days.

mlaventure commented:
@WeiZhang555 your change to engine-api needs to be vendored in.

thaJeztah commented:
@mlaventure @vdemeester already mentioned it :smile:

mlaventure commented:
Oups, sorry! :)

WeiZhang555 committed:
vendor engine-api 8924d6900370b4c7e7984b

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
Vendor has been done, PTAL :)

vdemeester commented:
@WeiZhang555 `\o/`

vdemeester commented:
LGTM 🐯

thaJeztah commented:
Thanks @WeiZhang555!
